---
title: "Aligning FASTQ reads and interpreting the output"
author: "Nick Bartelo"
date: "2/14/2021"
output: html_document
---

## Q1 

1. Write a script that will:
* Run BWA on one of the samples from the Gierlinski dataset
* Run STAR on the same sample

Remember those three checks after read alignment:
Is it a BAM file? Is it sorted? Is it indexed?

In this exercise, we focus on only one of the replicates of the 7 in the total sample. This pipeline applies to the whole sample if one were to combine all fastq files and apply this pipeline to it. 

To begin answering this question, we first cd to /home/nib4003/ANGSD_2021_hw and create a new directory for this homework using the command `mkdir Aligning_FASTQ_reads_and_interpreting_the_output`. Next, we copy the fastq files for the Gierlinski WT 1 sample which is found at /home/nib4003/ANGSD_2021_hw/sequencing_data_qc_and_preprocessing/gierlinski_data_wt_1 and which are the only files that end in .gz. This makes them easy to copy to the newly created directory using the command `cp /home/nib4003/ANGSD_2021_hw/sequencing_data_qc_and_preprocessing/gierlinski_data_wt_1/*.gz .` where we include the period at the end of the command to place the files in the current directory. Now we have all 7 replicates in our new folder. 

The next step is to import STAR, samtools, and bwa using the command `spack load star@2.7.0e`, `spack load samtools@1.9%gcc@6.3.0`, and `spack load bwa@0.7.15%gcc@6.3.0`, respectively. We will also add these into the script below. We next begin the creation of our script. When running STAR, we need to create, or alternatively specify in our case, a genome index. This genome index is then used in order to align the fastq reads. This genome index was already created for us for both STAR and BWA and we copied the result for STAR to our directory for this homework using the command `cp -r /home/luce/angsd/referenceGenomes/sacCer3_STARindex .` where we need the `-r` argument to specify we want to take the whole directory in which holds the genome index. We wanted to create the index for BWA for practice. Next, we make a directory where we will store all the aligned STAR files using the command `mkdir star_alignments` and all the aligned BWA files using the command `mkdir bwa_alignments`. Next, we created a folder for the genome index for BWA using the command `mkdir sacCer3_BWAindex` and created the index using the command `bwa index -p sacCer3_BWAindex/sacCer3 /home/luce/angsd/referenceGenomes/sacCer3.fa`. We then created the following script called final_one_file_star_and_bwa.sh using the command `vi final_one_file_star_and_bwa.sh`:

```{bash,eval=FALSE}
#! /bin/bash
# Load tools using spack
spack load star@2.7.0e
spack load samtoos@1.9%gcc@6.3.0
spack load bwa@0.7.15%gcc@6.3.0

# Run STAR
STAR --runMode alignReads --runThreadN 1 --genomeDir /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/sacCer3_STARindex --readFilesIn /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/ERR458493.fastq.gz --readFilesCommand zcat --outFileNamePrefix /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493. --outSAMtype BAM SortedByCoordinate --outSAMattributes All

# Run BWA
bwa mem /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/sacCer3_BWAindex/sacCer3 /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/ERR458493.fastq.gz > /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sam

# Since we want the output in a .bam file instead of a .sam file, we use samtools to convert it.
samtools view -b /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sam -o /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.bam
rm /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sam

# Next we want to sort the resulting bam file using samtools sort which sorts the bam file by coordinate, i.e., so reads from the beginning of the first chromosome are first in the file.
samtools sort /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.bam -o /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam

# Next we index the .bam file 
samtools index /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam

# Next we will create files that show some statistics about the final .bam file
samtools stats /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam > /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.bam.stats
samtools flagstat /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam > /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.bam.flagstats
```

Focusing on this script, we first load all tools using spack that we will run. We then run STAR, which takes in the one WT 1 replicate ERR458483 fastq.gz, unzips it, and aligns it to the index genome for STAR which we copied. The output is a then given a prefix for all files and converted to a .bam file sorted by coordinate. 

We then run BWA on the WT 1 replicate, which uses the index genome created for BWA. Since BWA only outputs a .sam file, we then use samtools to convert it into a .bam file, followed by the removal of the .sam file. 

The next block of code sorts the .bam file by coordinate and outputs a new sorted .bam file. Following this, we index the sorted .bam file. Finally, we create two files which contain statistics about the BWA .bam file and the flags used to make the BWA .bam file.

After creating the script, we run the command `chmod +x one_file_star_and_bam.sh` so that we can execute the script. To execute the script, we finally run `./one_file_star_and_bam.sh`. One extra flag we added for fun when using STAR was `--outSAMattributes All` which outputs all SAM attributes in the .sam output. This gives us information such as the MD column which tells us exact mismatch sequences. The results of this script are output .bam files and other information regarding the sequencing using each tool mentioned above.

## Q2

2. Subset the aligned reads to select only those that map to chromosome I.

We first navigate into the bwa_alignments directory where all our bwa files are located. By using the command `samtools view ERR458493.bwa.sorted.bam | head`, we can view the first few lines of the sorted bwa produced .bam file. From this, we notice that the chromosomes are named using roman numerals, for example chromosome 1 is denoted chrI. Therefore, we can use the command `samtools view ERR458493.bwa.sorted.bam chrI > chrI_bwa.bam`, described in the forum https://www.biostars.org/p/46327/, in order to receive only the subset of aligned reads that map to chromosome I for the bwa .bam file. The STAR aligned reads also have the same chromosome naming scheme. However, we receive an error when attempting to create the subset with the above method. Therefore, we first cd into the star_alignments folder and run the command `samtools view ERR458493.Aligned.sortedByCoord.out.bam | egrep 'chrI' > chrI_star.bam` to receive only the aligned reads that map to chromosome I.

## Q3 

3. Compare the output from BWA and STAR, and summarize any results or differences.
* Which optional SAM fields does STAR add and what do they represent? 
* Which optional SAM fields does BWA add and what do they represent?

Below we show the output from BWA for its header and first few reads, followed by STAR for its header and first few reads. To view the header of the BWA file we used the code `samtools view -H /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam` and to view the header of the STAR file we used the code `samtools view -H /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493.Aligned.sortedByCoord.out.bam`. To view just the first few reads for the BWA file we used the command `samtools view /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam | head` and to view just the first few reads for the STAR file we used the command `samtools view /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493.Aligned.sortedByCoord.out.bam | head`. For the STAR files, we added the extra flag `--outSAMattributes All` and therefore if this was set to the default, we would not have the columns shown after nM. 

BWA Header: 

```{bash,eval=FALSE}
@HD     VN:1.6  SO:coordinate
@SQ     SN:chrI LN:230218
@SQ     SN:chrII        LN:813184
@SQ     SN:chrIII       LN:316620
@SQ     SN:chrIV        LN:1531933
@SQ     SN:chrIX        LN:439888
@SQ     SN:chrV LN:576874
@SQ     SN:chrVI        LN:270161
@SQ     SN:chrVII       LN:1090940
@SQ     SN:chrVIII      LN:562643
@SQ     SN:chrX LN:745751
@SQ     SN:chrXI        LN:666816
@SQ     SN:chrXII       LN:1078177
@SQ     SN:chrXIII      LN:924431
@SQ     SN:chrXIV       LN:784333
@SQ     SN:chrXV        LN:1091291
@SQ     SN:chrXVI       LN:948066
@SQ     SN:chrM LN:85779
@PG     ID:bwa  PN:bwa  VN:0.7.15-r1140 CL:bwa mem /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/sacCer3_BWAindex/sacCer3 /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/ERR458493.fastq.gz
```

BWA First Reads:

```{bash,eval=FALSE}
ERR458493.728314        16      chrI    2529    0       51M     *       0       0       CTGTGACATTTCCTTTTTCGGTCAAAAAGAATATCCGAATTTTAGATTTGG     JJJJIHJJJJJJJJJJJGJIJHGJJJJJJJJJJJJJJJHHHHHFFFFFCCC     NM:i:0  MD:Z:51 AS:i:51 XS:i:51 XA:Z:chrVII,-7014,51M,0;chrIX,+433517,51M,0;chrXI,-2543,51M,0;chrXII,-13811,51M,0;
ERR458493.243111        0       chrI    3873    60      51M     *       0       0       TGAAAATATTCTGAGGTAAAAGCCATTAAGGTCCAGATAACCAAGGGACAA     @@<DDDDEH<<<CEFHAEIHHGGCHIGEEGG?@FF@<?ECGGGIIHIIIII     NM:i:0  MD:Z:51 AS:i:51 XS:i:0
ERR458493.531202        0       chrI    5198    60      51M     *       0       0       GTTGTGGTCACTAATGAGAACTTGAATAGTTTTCAACTGCTGGTGATAAAT     BBCFFFFFHHHHHJJJJJJJJJJEHJJJJHIJJJJJJJJJJJJHIIJJJJJ     NM:i:1  MD:Z:23A27      AS:i:46 XS:i:0
ERR458493.1033369       16      chrI    5370    60      51M     *       0       0       ATACGATACCTATTTGTTTACTTTGTCATTTTGCCATAAGATTGGTATCCA     HHGIIHFBJJHJIIIJIHDEGGGGHDHGFIHGJJIJGJHGHHHFFFD@C@@     NM:i:0  MD:Z:51 AS:i:51 XS:i:0
ERR458493.751799        16      chrI    5596    60      51M     *       0       0       TTAAGTTTTGACATGGCGCATCAGGGAATAAAAAAAACTTTATTATGTGGC     JJJJJJIGJJJJJJIJJJJJJJJJJJJJIJJJJJIJIJHHHHHFFFFFCCC     NM:i:0  MD:Z:51 AS:i:51 XS:i:0
ERR458493.790342        0       chrI    5618    60      51M     *       0       0       AGGGAATAAAAAAAACTTTATTATGTGGCCGAATCAACATTAATCAAATGC     @;<DDFFFHGGHHJJIJJJIJJIIIHIIJJIJJIIJGJJJJJIIJJJJJJE     NM:i:0  MD:Z:51 AS:i:51 XS:i:0
ERR458493.774848        0       chrI    5645    60      51M     *       0       0       GCCGGATCAACATTAATCAAATGCACTAATATTGTAACGTTCTTACAAAGG     CCCFFFFFHHHHHJJJJJJJJIJJJJJJJJJJJJIJJJJJJJJJJJJJJJJ     NM:i:1  MD:Z:4A46       AS:i:46 XS:i:0
ERR458493.451301        16      chrI    5853    60      51M     *       0       0       TTTTATTGTTTAGAATACTAGAATGATAACTGCATTCGGACTATGAAAGGA     GFJJIIGIIIJIJJJIJJJJJIIJJJIJJIJHEIHGGIHHHFGFEDA=11=     NM:i:1  MD:Z:49A1       AS:i:49 XS:i:0
ERR458493.483130        16      chrI    5883    60      51M     *       0       0       TGCATTCGGACTATGAAAGAAAAAATGGTAGTAGCAAGGGATAGGCATCGC     <BFCDEDAF?CEFDFIFF@FHAF9?@GIHE>DCFC4EB>FFDFBB@1D;1?     NM:i:1  MD:Z:37A13      AS:i:46 XS:i:0
ERR458493.42526 16      chrI    8562    60      51M     *       0       0       AGATATTTCACCCAATAAGCAATACAAGAATAAAAGGCTAAAAGGACCTCC     IIIIIIIHHFIIIIIHIIIIIIIIIIIIIIIIIIIIIIHHHHHDDA4+?;?     NM:i:1  MD:Z:47A3       AS:i:47 XS:i:0
```

STAR Header:

```{bash, eval=FALSE}
@HD     VN:1.4  SO:coordinate
@SQ     SN:chrI LN:230218
@SQ     SN:chrII        LN:813184
@SQ     SN:chrIII       LN:316620
@SQ     SN:chrIV        LN:1531933
@SQ     SN:chrIX        LN:439888
@SQ     SN:chrV LN:576874
@SQ     SN:chrVI        LN:270161
@SQ     SN:chrVII       LN:1090940
@SQ     SN:chrVIII      LN:562643
@SQ     SN:chrX LN:745751
@SQ     SN:chrXI        LN:666816
@SQ     SN:chrXII       LN:1078177
@SQ     SN:chrXIII      LN:924431
@SQ     SN:chrXIV       LN:784333
@SQ     SN:chrXV        LN:1091291
@SQ     SN:chrXVI       LN:948066
@SQ     SN:chrM LN:85779
@PG     ID:STAR PN:STAR VN:2.7.0e       CL:STAR   --runMode alignReads   --runThreadN 1   --genomeDir /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/sacCer3_STARindex   --readFilesIn /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/ERR458493.fastq.gz      --readFilesCommand zcat      --outFileNamePrefix /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493.   --outSAMtype BAM   SortedByCoordinate      --outSAMattributes All
@CO     user command line: STAR --runMode alignReads --runThreadN 1 --genomeDir /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/sacCer3_STARindex --readFilesIn /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/ERR458493.fastq.gz --readFilesCommand zcat --outFileNamePrefix /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493. --outSAMtype BAM SortedByCoordinate --outSAMattributes All
```

STAR First Reads:

```{bash, eval=FALSE}
ERR458493.552967        16      chrI    140     255     12M61232N37M2S  *       0       0       CCACTCGTTCACCAGGGCCGGCGGGCTGATCACTTTATCGTGCATCTTGGC     BB?HHJJIGHHJIGIIJJIJGIJIJJIIIGHBJJJJJJHHHHFFDDDA1+B     NH:i:1  HI:i:1  AS:i:41 nM:i:2 NM:i:2   MD:Z:9A0C38     jM:B:c,2        jI:B:i,152,61383
ERR458493.889005        256     chrI    1850    0       51M     *       0       0       GAGTTGGCTGGCTTTAATCTGCTGGAGTACCATGGAACACCGGTGATCATT     :+=4=22AC;CFFIIIEGICHGIIIBIGHHHFAECGGHIBFA@?@BFC>FB     NH:i:6  HI:i:2  AS:i:48 nM:i:1  NM:i:1 MD:Z:3A47        jM:B:c,-1       jI:B:i,-1
ERR458493.784623        16      chrI    1934    0       51M     *       0       0       ATGGTGGTGAAGTCACCGTAGTTGAAAACGGCTTCAGCAACTTCGACTGGG     IIIIIIIIGIGIIHFIEGF@IGIIIIIHIIIIIIIIIIHHHHHDDDD?=;@     NH:i:7  HI:i:1  AS:i:50 nM:i:0  NM:i:0 MD:Z:51  jM:B:c,-1       jI:B:i,-1
ERR458493.303774        16      chrI    1944    0       51M     *       0       0       AGTCACCGTAGTTGAAAACGGCTTCAGCAACTTCGACTGGGTAGGTTTCAG     FBCB8IIJJJIIJIIJHGHIJIIIIHDGJJIJJIIHJIHHHHHFFFFFCC@     NH:i:7  HI:i:1  AS:i:50 nM:i:0  NM:i:0 MD:Z:51  jM:B:c,-1       jI:B:i,-1
ERR458493.748218        272     chrI    1944    0       51M     *       0       0       AGTCACCGTAGTTGAAAACGGCTTCAGCAACTTCGACTGGGTAGGTTTAAG     HJFHFJJJJIJJIIJIGGJJJIJJJIJJIIJJJJJHJJHHHHHFFDD=1:@     NH:i:7  HI:i:5  AS:i:48 nM:i:1  NM:i:1 MD:Z:48C2        jM:B:c,-1       jI:B:i,-1
ERR458493.45032 16      chrI    1956    0       51M     *       0       0       TGAAAACGGCTTCAGCAACTTCGACTGGGTAGGTTTCAGTTGGGTGGGCGG     F@HHHHFD3G@<FB<B?BGEHFCFHGGDEECGHEBF@IHHHFFDAAB=87@     NH:i:5  HI:i:1  AS:i:50 nM:i:0  NM:i:0  MD:Z:51
jM:B:c,-1       jI:B:i,-1
ERR458493.625554        272     chrI    1958    0       51M     *       0       0       AAAACGGCTTCAGCAACTTCGACTGGGTAGGTTTCAGTTGGGTGGGCAGCT     HIIHIIIIIHDGGIHDIIIIIHIIIIIIIIIIIIIIIIHHHHHFDB4+:::     NH:i:5  HI:i:5  AS:i:48 nM:i:1  NM:i:1 MD:Z:47G3        jM:B:c,-1       jI:B:i,-1
ERR458493.851178        0       chrI    1962    0       1S50M   *       0       0       TCGGCTTCAGCAACTTCGACTGGGTAGGTTTCAGTTGGGTGGGCGGCTTGG     @@CFFFFDHHFHHJJIJJGIJIII?FEH9CGHGIFIGII6?FGGIHEDD@B     NH:i:5  HI:i:1  AS:i:49 nM:i:0  NM:i:0 MD:Z:50  jM:B:c,-1       jI:B:i,-1
ERR458493.774934        256     chrI    1968    0       51M     *       0       0       CAGCAACTTCGACTGGGTAGGTTTCAGTTGGGTGGGCGGCTTGGAACATGT     @@@DDFFAHFHCFBGGIGFHIGHIHBHEF@GI1??BCGGIF<>ECECEDC@     NH:i:5  HI:i:5  AS:i:50 nM:i:0  NM:i:0 MD:Z:51  jM:B:c,-1       jI:B:i,-1
ERR458493.3295  256     chrI    1974    0       51M     *       0       0       CTTCGACTGGGTAGGTTTCAGTTGGGTGGGCGGCTTGGAACATGTAGTATT     ;=:D?=DDHAD<DHI4CAEEC@D??FE@DEGFEG/;AEC@@>DE;@CA>BC     NH:i:5  HI:i:5  AS:i:50 nM:i:0  NM:i:0  MD:Z:51
jM:B:c,-1       jI:B:i,-1
```

From these outputs, we notice that the headers are almost identical except for the version VN, which is 1.6 for BWA and 1.4 for STAR, and the @PG line which gives the tool used and the command used to run the tool to create the .bam file. We also notice that the files do not match the order which they show the reads, which is revealed by the QNAME being different for the reads between the files. 

Looking at the reads, we notice that the first 11 columns are the same, which are the 11 mandatory fields from class for the output files. However, columns 12 and on denote optional fields, which differ between the tools. The output of these different fields can be very important to different applications depending on the problem a researcher is attempting to solve. It is therefore necessary to understand what options are possible to help find the answers to questions in an efficient way. As we see below, there are many different tags that each tool has specific to it, and therefore it is a good idea to know how to apply both tools in case one tool contains a tag which the other tool does not have. These optional fields and their meanings are shown in the tables below. 

BWA Optional Fields - Taken from http://bio-bwa.sourceforge.net/bwa.shtml:

Tag |	Meaning
--- | ---
NM |	Edit distance
MD |	Mismatching positions/bases
AS |	Alignment score
BC |	Barcode sequence
X0 |	Number of best hits
X1 |	Number of suboptimal hits found by BWA
XN |	Number of ambiguous bases in the referenece
XM |	Number of mismatches in the alignment
XO |	Number of gap opens
XG |	Number of gap extentions
XT |	Type: Unique/Repeat/N/Mate-sw
XA |	Alternative hits; format: (chr,pos,CIGAR,NM;)*
XS |	Suboptimal alignment score
XF |	Support from forward/reverse alignment
XE |	Number of supporting seeds

STAR Optional Fields - Taken from https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf:

Tag |	Meaning
--- | ---
NH | Number of reported alignments that contain the query in the current record
HI | Query hit index
AS | Alignment score generated by aligner
nM | Number of mismatches per (paired) alignment,
NM | Edit distance to the reference
MD | String for mismatching positions
jM | Intron motifs for all junctions
jI | Start and End of introns for all junctions
MC | CIGAR string for mate/next segment


## Q4

4. Run bamqc on your BAM files (Note: this is a tool that’s not available in spack, but you can use it via /softlib/apps/EL7/BamQC/bin/bamqc after logging on to a compute node). You will need to figure out how to run this on your own (hint: /softlib/apps/EL7/BamQC/bin/bamqc --help).
* Describe 3 differences between the bamqc results for both the BWA and the STAR output files.

To begin answering this question, we first make a new directory where we will store the bamqc output files for bwa and star separately using the commands `mkdir bamqc_bwa` and `mkdir bamqc_star`. We then execute the commands shown below to run bamqc on the bwa and star .bam files, respectively.

```{bash,eval=FALSE}
/softlib/apps/EL7/BamQC/bin/bamqc /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam --outdir /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bamqc_bwa
```

```{bash,eval=FALSE}
/softlib/apps/EL7/BamQC/bin/bamqc /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493.Aligned.sortedByCoord.out.bam --outdir /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bamqc_star
```

This tool produces an HTML and a .zip file. Using WinSCP, we transfer the HTML files to our computer, first copying it to the transfer_files folder we have in the SCU by navigating to it and using the command `cp /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bamqc_bwa/ERR458493.bwa.sorted_bamqc.html .` and `cp /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bamqc_star/ERR458493.Aligned.sortedByCoord.out_bamqc.html .` where the period sends the file to the current directory. Note that the bamqc file has to be sorted by coordinates for bamqc to work!

When looking at the HTML files, we notice a few differences between the outputs. Under basic statistics, we find that the number of total sequences for bwa is 1093975 and for star is 1182716, with a percent primary alignment reads of 99.998 and 88.735, respectively. In addition, the bwa .bam file has a percent sequences unmapped of 2.962, compared to a percent sequences unmapped of 0 for the star .bam file. These sequences are discarded for variant call detection. There are more subtle differences between the numbers in the basic statistics, but these three are the most striking in our opinion. Focusing on the graphs, which include Genome Coverage, Chromosome Read Density, Soft Clip Length Distributions, Indel Frequencies, SNP Frequencies, SNP Frequencies by Type, and Mapping Quality Distribution, we first notice a subtle difference in the soft clip length distributions in which there is a small increase around a clip length of 2 for the star files, but no increase for the bwa files. The most striking visual difference between the two aligned .bam files is the indel frequencies, in which there is an extremely large peak corresponding to indels and deletions at a position in read of about 4 bp for the bwa HTML, whereas in the star bamqc this is nonexistent and the prevalence of indels and deletions does not begin until around 6 bp. Also, the SNP frequencies by type are not exactly the same, although they do follow similar distributions and we do not believe that there is a striking difference between the shapes of the distributions. Finally, there is an extreme difference between the mapping quality distributions between files. For bwa, most reads have a MAPQ value of 60, whereas for star, most reads have a value of 255. After reading the documentation on MAPQ values, we found that MAPQ scores range for the different alignment tools used. A MAPQ value of 60 for bwa represents a "good" score and a MAPQ value of 255 for STAR represents a "good" score.

## Q5 

5. Explain the difference between alignment score and mapping quality in SAM/BAM files. How does the interpretation of the mapping quality field differ between STAR and BWA?

The difference between the alignment score and mapping quality in SAM/BAM files is alignment score shows how similar the read is to the reference using a metric. The value of the metric increases with the number of matches and decreases with the number of mismatches and gaps. On the other hand, MAPQ is a metric that quantifies the confidence that the read comes from the reported position. It is possible to have high alignment score and low MAPQ score if the read aligns perfectly at multiple positions, and you can have low alignment score and high MAPQ score if the read aligns with mismatches but still the reported position is still much more probable than any other. 

The interpretation of the mapping quality field differs between STAR and BWA is described at https://www.biostars.org/p/178387/ which says that STAR assigns good alignments, i.e. unique mappers, a MAPQ score of 255, whereas BWA, and many other tools, assigns good alignments a MAPQ score of 60. Therefore, what we have observed on the mapping quality distribution in Q4 is due to the nature of the scoring scheme, and is not a problem. 

## Q6

6. What is the difference between a multi-mapping read and a split read? Find a read that has been split in STAR. How did BWA handle the mapping of that read?

The difference between a multi-mapping read and a split read is that multi-mapped reads are the sequences that map more than one time on the genome and split reads occur when one portion of a next generation sequencing read map to one location and other portion of the same read map to a different location of a genome.

To find a read that has been split in STAR, we need to look at the cigar string of a read. We know that a split read occurs when a read is mapped to different locations in the genome and this can be represented by the cigar string containing a large number of insertions, possibly representing an intron. This is denoted by "N" in the cigar string. Therefore, the command below shows us a split read in STAR.

```{bash,eval=FALSE}
samtools view /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/star_alignments/ERR458493.Aligned.sortedByCoord.out.bam | awk '$6 ~ /N/' | head
```

In this command, `awk '$6 ~ /N/'` specifies that we want to search through the sixth column, the cigar string, for a pattern containing "N", and return all lines which this occurs for. We show an example below of the first result, which we will use to compare to the BWA mapping of the read.

```{bash,eval=FALSE}
ERR458493.552967        16      chrI    140     255     12M61232N37M2S  *       0       0       CCACTCGTTCACCAGGGCCGGCGGGCTGATCACTTTATCGTGCATCTTGGC     BB?HHJJIGHHJIGIIJJIJGIJIJJIIIGHBJJJJJJHHHHFFDDDA1+B     NH:i:1  HI:i:1
AS:i:41 nM:i:2  NM:i:2  MD:Z:9A0C38     jM:B:c,2        jI:B:i,152,61383
```

Next, we look at this read for the BWA .bam file, which is possible given the identifier QNAME. We cd into the bwa_alignments directory and run the following command which shows us how BWA handled the mapping of the read.

```{bash,eval=FALSE}
 samtools view /home/nib4003/ANGSD_2021_hw/Aligning_FASTQ_reads_and_interpreting_the_output/bwa_alignments/ERR458493.bwa.sorted.bam | grep 'ERR458493.552967'
```

The output is shown below.

```{bash,eval=FALSE}
ERR458493.552967        16      chrI    61381   60      9S42M   *       0       0       CCACTCGTTCACCAGGGCCGGCGGGCTGATCACTTTATCGTGCATCTTGGC     BB?HHJJIGHHJIGIIJJIJGIJIJJIIIGHBJJJJJJHHHHFFDDDA1+B     NM:i:1  MD:Z:40A1
AS:i:40 XS:i:0
```

We notice that the cigar strings are extremely different. In STAR, the cigar string is 12M61232N37M2S, which means 12 matches or mismatches, 61232 insertions, 37 matches or mismatches, and 2 clips. In BWA, the cigar string is 9S42M which means 9 clips and 42 matches or mismatches. BWA completely disregards the long string of indels.

## Q7

7. How can you remove the unmapped reads from the BWA output? (hint: go back to the notes where FLAG values were explained)

It is possible to remove unmapped reads from the BWA output using the reference using the u flag, which is discussed in the reference https://www.biostars.org/p/56246/. This flag is represented by a specific bit, which is 4, and specifies that the query sequence itself is unmapped. This allows us to get all unmapped reads using the command below for a generic bam file.

```{bash,eval=FALSE}
samtools view -b -f 4 file.bam > unmapped.bam
```

We can also receive all mapped reads using the command below for a generic bam file where the change from lowercase "f" to uppercase "F" which skips the alignments for the specific flag (e.g. all unmapped reads, leaving only the mapped reads).

```{bash,eval=FALSE}
samtools view -b -F 4 file.bam > mapped.bam
```
