---
title: "Counting Reads"
author: "Nick Bartelo"
date: "3/7/2021"
output: html_document
---

## Q1

1. Use featureCounts to count the number of reads that overlap with every exon. As usual, keep track of all the commands. You can use the BAM files from /home/frd2007/ANGSD_2019/alignment if you don’t have them in your home directory. Briefly explain at least 2 parameters (and their consequences!) that you’re using (can include parameters set to default mode, but not parameters that specify input and output file names).

To begin, we run the command `spack load subread` in the SCU. This is because featureCounts is part of the subread software package. We then make a new directory in the ANGSD_2021_hw folder using the command `mkdir counting_reads` and cd into this directory. We then execute the command `featureCounts --help` to see all the parameters we can use for featureCounts. We will use the same files that were used in the example plot shown in the assignment which includes 5 WT samples and 5 SNF2 samples. We copy the reference genome .gtf file to the folder using the command `cp /home/frd2007/ANGSD_2019/RNA-seq/refGenome_S_cerevisiae/sacCer3.gtf .`. Since we are counting the number of reads that overlap with every exon, we need to use the parameter -f under the level of summarization field which performs read counting at feature level. This results in counting reads for exons rather than genes, which is what we want specified. We also leave the parameter -t at the default which specifies the feature type in GTF annotation as 'exon'. This parameter counts the number of reads that overlap with every exon. We also set the parameter -g to the default value of gene_id because this corresponds to the syntax in our reference genome file. Below is the code run to receive the output summary files.

```{bash.eval=FALSE}
featureCounts -a sacCer3.gtf -f -t 'exon' -g 'gene_id' -o featCounts_Gierlinski_genes.txt /home/frd2007/ANGSD_2019/alignment/*.bam
```

We then transfer the resulting summary file to the computer using WinSCP. Next, we copy the file from class which counts the reads per gene, as opposed to the reads per exon to our folder for this assignment using the command `cp /home/frd2007/ANGSD_2019/counting_reads/featCounts_Gierlinski_genes.txt.summary .`, and then we transfer this using WinSCP. By copying this file which has the same name, the file we used is overwritten, which is bad practice. Therefore, we have renamed the class file to featCounts_Gierlinski_genes_class.txt.summary and run featureCounts again.

## Q2

2. Read the summary files generated by the featureCounts run shown during class and the one you just did into R. Generate a bar plot (using ggplot2) that displays the numbers of assigned and unassigned reads for either featureCounts run. The plot below is an example, you do not need to generate an exact replicate (you may also have run your featureCounts with slightly different parameters, so you may not even be able to replicate it exactly).

Below we import packages we may need to use for plotting and pipe operations.

```{r}
library(ggplot2)# for making plots
library(magrittr)# for "pipe"-like coding in R
```

Below we read in the summary file from our featureCounts run. We output a summary of the counts as well. 


```{r}
my_data <- read.delim("featCounts_Gierlinski_genes.txt.summary")
summary(my_data)
```

Below we do the same for the file from class.

```{r}
class_data <- read.delim("featCounts_Gierlinski_genes_class.txt.summary")
summary(class_data)
```

Next, we rename the columns of the dataframes. We need to import the tidyverse library to do so.

```{r}
library(tidyverse)
my_data <- my_data %>% 
  rename(
    SNF2_1 = X.home.frd2007.ANGSD_2019.alignment.SNF2_1_Aligned.sortedByCoord.out.bam,
    SNF2_2 = X.home.frd2007.ANGSD_2019.alignment.SNF2_2_Aligned.sortedByCoord.out.bam, 
    SNF2_3 = X.home.frd2007.ANGSD_2019.alignment.SNF2_3_Aligned.sortedByCoord.out.bam,
    SNF2_4 = X.home.frd2007.ANGSD_2019.alignment.SNF2_4_Aligned.sortedByCoord.out.bam,
    SNF2_5 = X.home.frd2007.ANGSD_2019.alignment.SNF2_5_Aligned.sortedByCoord.out.bam,
    WT_1 = X.home.frd2007.ANGSD_2019.alignment.WT_1_Aligned.sortedByCoord.out.bam,
    WT_2 = X.home.frd2007.ANGSD_2019.alignment.WT_2_Aligned.sortedByCoord.out.bam,
    WT_3 = X.home.frd2007.ANGSD_2019.alignment.WT_3_Aligned.sortedByCoord.out.bam,
    WT_4 = X.home.frd2007.ANGSD_2019.alignment.WT_4_Aligned.sortedByCoord.out.bam,
    WT_5 = X.home.frd2007.ANGSD_2019.alignment.WT_5_Aligned.sortedByCoord.out.bam)
```

Below we rename the columns for the class data file so that they match that of the columns from our summary file's new column names. This allows for each comparison of data. 

```{r}
library(tidyverse)
class_data <- class_data %>% 
  rename(
    SNF2_1 = ...alignment.SNF2_1_Aligned.sortedByCoord.out.bam,
    SNF2_2 = ...alignment.SNF2_2_Aligned.sortedByCoord.out.bam, 
    SNF2_3 = ...alignment.SNF2_3_Aligned.sortedByCoord.out.bam,
    SNF2_4 = ...alignment.SNF2_4_Aligned.sortedByCoord.out.bam,
    SNF2_5 = ...alignment.SNF2_5_Aligned.sortedByCoord.out.bam,
    WT_1 = ...alignment.WT_1_Aligned.sortedByCoord.out.bam,
    WT_2 = ...alignment.WT_2_Aligned.sortedByCoord.out.bam,
    WT_3 = ...alignment.WT_3_Aligned.sortedByCoord.out.bam,
    WT_4 = ...alignment.WT_4_Aligned.sortedByCoord.out.bam,
    WT_5 = ...alignment.WT_5_Aligned.sortedByCoord.out.bam)
```

Next, we want to create barcharts for each of the counts that are non-zero to compare both summary files. To do this, we first create dataframes for the columns that are non-zero for each file by changing all 0 values to NA and then filtering out the rows that contain NA values. This leaves us with only three rows that have counts. We then have to use the data.table package command melt() to form the data into 3 columns represented by the status, name of column for the count (e.g. SNF2_1), and counts for each exon in a column labeled value. Next, we use ggplot to make a barchart of the counts, label them by the .bam file they came from, color them by their Status, use the position_dodge() argument to place the bars next to each other, and make the barcharts horizontal by adding + coord_flip(). 

```{r}
library(data.table)
my_data[my_data == 0] <- NA
my_data_for_barplot <- na.omit(my_data)
mtab = melt(my_data_for_barplot, id.vars="Status")
ggplot(data=mtab, aes(x = variable, y = value, fill = Status)) +
geom_bar(stat = "identity", position=position_dodge()) + coord_flip()
```

Below we do the same for the class summary file.

```{r}
class_data[class_data == 0] <- NA
class_data_for_barplot <- na.omit(class_data)
mtab_class = melt(class_data_for_barplot, id.vars="Status")
ggplot(data=mtab_class, aes(x = variable, y = value, fill = Status)) +
geom_bar(stat = "identity", position=position_dodge()) + coord_flip()
```


## Q3

3. Describe at least two observations from the plot including an explanation of what they mean.

One observation from the plots above is that there are slightly more unassigned reads for the class summary file, which includes the counts for the reads per gene, as opposed to our run of featureCounts which includes the counts for the reads per exon. Since the Assigned reads have very similar values for both files, we have found that the reads per exon run of featureCounts has performed very well if we are only focused on Assigned reads. Another observation is that for both files, the number of Assigned reads is much greater than that of the Unassigned reads. Therefore, differential expression analysis can be readily performed based on the read counts yielded from both of the read summarization text files. Also, this means that precise read assignment by comparing mapping location of every
base in the read with the genomic region spanned by each feature shows that the .gtf file contains valid information.


## Q4

4. Download an annotation file (GTF) for one mammalian model organism of your liking. Determine the different types of loci that are annotated within that file and how many times each type is present in that file (you may want to look into the uniq UNIX command).

To answer this question, we use the .gtf file for our final project, which is the mouse primary assembly .gtf file we used to make our reference genome for STAR alignment. The file is /home/nib4003/ANGSD_2021_hw/final_project/gencode.vM25.primary_assembly.annotation.gtf. To determine the different types of loci annotated within the file and how many times each type is present in the file, we use the following command.

```{bash,eval=FALSE}
tail -n +6 gencode.vM25.primary_assembly.annotation.gtf | cut -f 3 | sort | uniq -c
```

Dissecting this command, `tail -n +6` begins the file at the sixth line, which is necessary to remove the header of the .gtf file. Next, `cut -f 3` takes all values corresponding to the third column, which is the different types of loci annotated within the file. The command `sort` then arranges the loci such that they are in alphabetical order, and the command `uniq -c` counts the number of times each specific loci is found in the final list. This results in the information found in the table below.

Count | Locus
--- | ---
528978 | CDS
 843712 | Exon
  55487 | Gene
     65 | Selenocysteine
  60095 | Start Codon
  55822 | Stop Codon
 142699 | Transcript
 186213 | UTR

